### PR TITLE
WT-6858 Initial implementation for hs cursor insert

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

**Review by hbokhari — APPROVED** [review]
LGTM!

File: src/cursor/cur_hs.c [hunk 1]
```diff
@@ -185,6 +184,151 @@ __curhs_reset(WT_CURSOR *cursor)
...
+
+    CURSOR_API_CALL_PREPARE_ALLOWED(cursor, session, insert, CUR2BT(file_cursor));
+
+    /* Allocate a tombstone only when there is a valid stop time point. */
+    if (WT_TIME_WINDOW_HAS_STOP(&hs_cursor->time_window)) {
+        /*
+         * Insert a delete record to represent stop time point for the actual record to be inserted.
+         * Set the stop time point as the commit time point of the history store delete record.
+         */
+        WT_ERR(__wt_upd_alloc_tombstone(session, &hs_upd, NULL));
```
Comment (bvpvamsikrishna) [inline src/cursor/cur_hs.c:307]: Don't we need to free this memory in the error handling of the function?

File: src/cursor/cur_hs.c [hunk 2]
```diff
@@ -185,6 +184,151 @@ __curhs_reset(WT_CURSOR *cursor)
...
+        hs_upd->start_ts = hs_cursor->time_window.stop_ts;
+        hs_upd->durable_ts = hs_cursor->time_window.durable_stop_ts;
+        hs_upd->txnid = hs_cursor->time_window.stop_txn;
+    }
+
+    /*
+     * Append to the delete record, the actual record to be inserted into the history store. Set the
+     * current update start time point as the commit time point to the history store record.
+     */
+    WT_ERR(__wt_upd_alloc(session, &file_cursor->value, WT_UPDATE_STANDARD, &upd_local, NULL));
```
Comment (bvpvamsikrishna) [inline src/cursor/cur_hs.c:317]: Don't we need to free this memory in the error handling of the function?

File: src/cursor/cur_hs.c [hunk 3]
```diff
@@ -185,6 +184,151 @@ __curhs_reset(WT_CURSOR *cursor)
...
+    CURSOR_API_CALL_PREPARE_ALLOWED(cursor, session, insert, CUR2BT(file_cursor));
+
+    /* Allocate a tombstone only when there is a valid stop time point. */
+    if (WT_TIME_WINDOW_HAS_STOP(&hs_cursor->time_window)) {
+        /*
+         * Insert a delete record to represent stop time point for the actual record to be inserted.
+         * Set the stop time point as the commit time point of the history store delete record.
+         */
+        WT_ERR(__wt_upd_alloc_tombstone(session, &hs_upd, NULL));
+        hs_upd->start_ts = hs_cursor->time_window.stop_ts;
```
Comment (bvpvamsikrishna) [inline src/cursor/cur_hs.c:308]: could we rename hs_upd to hs_tombstone.

File: src/cursor/cur_hs.c [hunk 4]
```diff
@@ -185,6 +184,151 @@ __curhs_reset(WT_CURSOR *cursor)
...
+        hs_upd->durable_ts = hs_cursor->time_window.durable_stop_ts;
+        hs_upd->txnid = hs_cursor->time_window.stop_txn;
+    }
+
+    /*
+     * Append to the delete record, the actual record to be inserted into the history store. Set the
+     * current update start time point as the commit time point to the history store record.
+     */
+    WT_ERR(__wt_upd_alloc(session, &file_cursor->value, WT_UPDATE_STANDARD, &upd_local, NULL));
+    upd_local->start_ts = hs_cursor->time_window.start_ts;
```
Comment (bvpvamsikrishna) [inline src/cursor/cur_hs.c:318]: could we rename upd_local to hs_upd?

**Review by bvpvamsikrishna — APPROVED** [review]
lgtm after changing the variable names

File: src/cursor/cur_hs.c (hunk 1 above)
Comment (quchenhao) [reply src/cursor/cur_hs.c]: fixed.

File: src/cursor/cur_hs.c (hunk 2 above)
Comment (quchenhao) [reply src/cursor/cur_hs.c]: fixed.

File: src/cursor/cur_hs.c (hunk 3 above)
Comment (quchenhao) [reply src/cursor/cur_hs.c]: fixed.

File: src/cursor/cur_hs.c (hunk 4 above)
Comment (quchenhao) [reply src/cursor/cur_hs.c]: fixed.

**Pull request merged**